### PR TITLE
Mk release location

### PR DESCRIPTION
**Pull request by mkangia**
[Spec](https://docs.google.com/document/d/1xfOaDsV0gllRwZM2zWigc4BEHarDjPKuwbGYp50-Xt0/edit#bookmark=id.1uejjreadd8r)

[Ticket](https://dimagi-dev.atlassian.net/browse/ICDS-301)

@esoergel Similar to what you saw in https://github.com/dimagi/commcare-hq/pull/22736 in case you are up to reviewing this

To Reviewer,
1. best review by commit but all at once should be easy too
2. instead of adding a new ff i used the same one as earlier and updated its description. Its used by icds only
3. A little preview of the UI [here](http://recordit.co/QOirWUfdDe), please ignore the kafka error in that demo. This demonstrates the view, the error message in case the app version is not present and how one can remove restriction by submitting form with empty value for version.

### CODE REVIEW

File: corehq/apps/app_manager/decorators.py [hunk 1]
```diff
@@ -70,8 +75,14 @@ def _safe_cached_download(request, *args, **kwargs):
             request.GET.pop('username')
 
         latest_enabled_build = None
-        if latest and request.GET.get('profile') and toggles.RELEASE_BUILDS_PER_PROFILE.enabled(domain):
-            latest_enabled_build = get_latest_enabled_build_for_profile(domain, request.GET.get('profile'))
+        if latest and toggles.RELEASE_BUILDS_PER_PROFILE.enabled(domain):
+            if request.GET.get('user_id', '6a31cb9e30bb4202ab199cad63d81be7'):
+                location_id = get_user_location_id(request.GET.get('user_id', '6a31cb9e30bb4202ab199cad63d81be7'))
```
Comment (mkangia) [inline corehq/apps/app_manager/decorators.py]: the id `6a31cb9e30bb4202ab199cad63d81be7` is removed later.

**Review by esoergel — COMMENTED** [review]
I'm not crazy about this approach.  It feels very weird for this to be part of locations, given that it's a feature that governs app releases.  Presumably, app managers will be using the feature, but it's on individual location pages, accessible to users who can edit locations.  This also means that it has a broad footprint across core locations code.

Also, now there are three separate data models controlling which app a user sees, and the interaction between them isn't obvious.  If someone needs to answer the question "why does/doesn't this user see this app version?"  The answer could be on this tab on that user's location details page, on the page from the profile-based releases feature, or on the releases page.  What if they release the app for that user's profile, but their location is still restricted, or vice-versa?

An alternative approach would be to have enabled profiles and locations as properties of the build (defaulting to "all").  Then users would receive the latest starred build that matches their location and profile.  The UI could also piggy-back on the existing releases page - select locations and select profiles.  You could also make the case that this is a complex enough release management workflow that it deserves its own UI.  Either way I strongly believe there should be one place that controls this.

Lastly, this doesn't take into account the hierarchical nature of locations.  If a build is released for a state, only users assigned directly to that state will receive it, those assigned to a district won't.

**Comment by snopoke** [comment]
I agree with @esoergel's point which is similar to what was originally specified in the spec for profile restrictions (storing this information on the build itself vs on the location or in a separate model).

One point to mention is that this would supersede the 'profile releases' which would be deprecated and removed in future.

**Comment by mkangia** [comment]
Thanks for the inputs @esoergel  @snopoke 

In regards to two different models to maintain (`LatestEnabledBuildProfiles` and `SQLLocation`) and then profile based control to be deprecated, i was checking with App builders and i got a push to keep both for now. I think they would majorly use the profile base restriction and use the location one for just one specific use case.

I agree on keeping this at a single place and will try to work towards it.
Though i am not in favour of keeping these saved on the build doc. If i save it on a build doc, it would duplicate in all builds i.e the information would then need to be saved on all successive builds even if it is not something that needs to be maintained as versions.
So i am going to work with the same interface "view source files"/add a new one and use the same model `LatestEnabledBuildProfiles` to keep logic at a single place.

let me know your thoughts.

**Comment by esoergel** [comment]
I imagine there's a lot of utility to have this sort of fine-grained control of rollout.  Makes sense that they may want both.

> Though i am not in favour of keeping these saved on the build doc. If i save it on a build doc, it would duplicate in all builds i.e the information would then need to be saved on all successive builds even if it is not something that needs to be maintained as versions.

What do you mean?  Isn't this very much build-specific?  I'd imagine something like:

* Version 1: Released for all profiles and locations
* Version 2: Released only for Uttar Pradesh
* Version 3: Released only for Bihar
* Version 4: Released for all profiles and locations

If we find a critical bug in v4 and it's unstarred, UP users should get v2, Bihar users should get v3, and all others should get v1.  When an app manager declares that Uttar Pradesh users see v4, that doesn't mean that everything before that is also a valid build for them.

That is, this information _does_ seem to me like something specific to each build, so it wouldn't be duplicated information.

File: corehq/apps/locations/forms.py [hunk 2]
```diff
@@ -539,6 +542,59 @@ def _get_custom_user_data(self, bound_data):
...
+
+    def save(self):
+        version = self.cleaned_data['version']
+        app_id = self.cleaned_data['app_id']
+        save = False
+        if app_id and not version and app_id in self.location.restricted_app_releases:
+            self.location.restricted_app_releases.pop(app_id)
+            save = True
+        elif app_id and version:
+            self.location.restricted_app_releases[app_id] = version
```
Comment (snopoke) [inline corehq/apps/locations/forms.py]: any reason not to store the build ID alongside the version? that would prevent having to look it up each time.

File: corehq/apps/app_manager/util.py [hunk 3]
```diff
@@ -692,6 +694,23 @@ def get_latest_enabled_build_for_profile(domain, profile_id):
...
+    from corehq.apps.app_manager.models import Application
+    # if its a build id, get the application id
+    app_or_release = Application.get_db().get(app_or_build_id)
+    return app_or_release.get('copy_of', app_or_build_id)
+
+
+@quickcache(['domain', 'location_id', 'app_id'], timeout=24 * 60 * 60)
+def get_latest_enabled_build_for_location(domain, location_id, app_id):
+    location = SQLLocation.active_objects.get(location_id=location_id)
+    version_to_update = location.restricted_app_releases.get(app_id)
```
Comment (snopoke) [inline corehq/apps/app_manager/util.py]: how will this work if the user's location is lower in the location hierarchy? e.g. this is a District but user location is AWC

**Comment by snopoke** [comment]
Again agreeing with Ethan. Storing the info on the build makes sense to me because then there is a single place to look to know who should be allowed to install it. I don't see why you would have to duplicate any info since you could query a couch view with the location / profile in the key which would give you the latest version for that location / profile.

The only challenge I see would be if the build is released for a location higher up the location hierarchy than the user's. E.g. build released for State A - user assigned to AWC 1. In this case you might have to query for each location in the hierarchy of the user.

**Comment by mkangia** [comment]
Synced up with app builders over last couple of days to finalise
1. They would switch completely to location based restriction. But requested to keep the code for profile based around temporarily in case things go wrong with location
2. in short, in case a location is restricted to a version, any child location should have the same restriction

**Comment by mkangia** [comment]
Closing this in the favour of a better PR.

**Pull request closed without merging**